Add session key permission provider

alexklbuckley opened the pull request:
Sponsored-by: Auckland University of Technology, New Zealand

TODO
- [x] Update changelog when merging (note factory changes)

alexklbuckley committed:
Add secure cookie permission provider

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
Hi @demiankatz , this is the follow-up secure cookie permission provider PR. It is based on the notes from https://github.com/vufind-org/vufind/pull/3828#issuecomment-2255721022 

Would you mind please taking a look? I would really appreciate any advice on the next steps for getting the secure cookie to work.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -49,7 +49,8 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
         'ipRange' => IpRange::class,
         'ipRegEx' => IpRegEx::class,
-        'role' => Role::class,
+    'role' => Role::class,
```
The indentation looks weird here. Is it possible your editor is using tabs instead of spaces somewhere?

alexklbuckley replied:
thanks @demiankatz that is fixed in my follow-up

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,79 @@
+<?php
+
+/**
+ * Secure cookie permission provider for VuFind.
```
I don't think you need to say "for VuFind" in the comment here. It's implied by the context!

alexklbuckley replied:
thanks @demiankatz that is removed in my follow-up

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,79 @@
+use VuFind\Cookie\CookieManager;
+
+/**
+ * Secure cookie permission provider for VuFind.
```
Ditto.

alexklbuckley replied:
thanks @demiankatz that is removed in my follow-up

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,79 @@
+     */
+    public function __construct(
+        protected CookieManager $cookieManager
+    ) {
```
My suggestion would be to add a second constructor parameter that's a `\Laminas\Session\Container` object (you can find many examples of building containers elsewhere in the code if you need them -- note that it's important in the factory to inject the session manager into the container when you create it, so you should copy this logic from another factory that creates containers). The container namespace could be something like 'SecureCookie'.

My thought is that when you set a cookie designed to authorize the user with this plugin, you set the cookie to a randomly-generated value like a uuid and then also store that value in the session container. Then when the cookie is validated, we confirm that the value in the session matches the value in the cookie, preventing remote spoofing since there is no way to predict the value of the cookie. We could create a public helper method on the provider to set the secret, and then any code setting an authorization cookie could fetch the provider out of the plugin manager and store the secret so that it will work as expected on subsequent requests.

Does that make any sense?

alexklbuckley replied:
Thanks @demiankatz ! Yes, I think so.  I've been working on a follow-up adding this, I will push it to this PR in the next day or two.

alexklbuckley replied:
thanks @demiankatz 

I've added a follow-up which hopefully does the following correctly:

- Add a second constructor parameter to SecureCookie.php which is a \Laminas\Session\Container object 

- In the SecureCookieFactory.php inject the session manager into the container when we create it

alexklbuckley replied:
Hi @demiankatz I have added the public helper method in the SecureCookie.php to set the secret (a random uuid) in the cookie and session container. 

I'm also validating in the SecureCookie.php if the value in the session matches the value in the cookie.

I'm not sure if I've done this right, so if you could you please advise what I should improve upon that would be much appreciated!

demiankatz reviewed (changes requested):
Thanks for getting this started, @alexklbuckley -- I've shared some ideas and comments below.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into secure_cookie_permission_provider

alexklbuckley committed:
Add secure cookie permission provider: Implement review feedback

- Fix indentation.
- Remove redundant code comments.

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -65,7 +66,8 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
         IpRange::class => IpRangeFactory::class,
         IpRegEx::class => IpRegExFactory::class,
-        Role::class => \Laminas\ServiceManager\Factory\InvokableFactory::class,
+    Role::class => \Laminas\ServiceManager\Factory\InvokableFactory::class,
```
There seem to be whitespace problems down here too...

alexklbuckley replied:
thanks @demiankatz I've included this in my follow-up

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley, I'll look forward to further progress as time permits. In the meantime, one more formatting issue.

alexklbuckley committed:
Add secure cookie permission provider: Implement review feedback

- Fix indentation.
- Add a second constructor parameter to SecureCookie.php that is a \Laminas\Session\Container object
- Inject the session manager into the container from
  SecureCookieFactory.php

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Add secure cookie permission provider: Implement review feedback

- Implement public helper function in SecureCookie.php to set a secret (a random UUID) in
  cookie and session container.
- In SecureCookie.php getPermissions() we validate if the random UUID value in the session
  matches the value in the cookie.

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return array
+     */
+    public function getPermissions($options)
```
If you want this to work similarly to InsecureCookie, then `$options` should be a string or array of cookie names. Right now this method seems to be hard-coded to use a cookie named uuid, but I think you should actually be checking based on the $options value(s).

Also note that we should add an entry to permissions.ini describing the new provider so the documentation is complete.

alexklbuckley replied:
thanks @demiankatz I have made those changes to the getPermissions() function in SecureCookie.php

I've added added an entry to permissions.ini documenting this new secure cookie provider.

Please let me know if you think the documented description needs clarifying/changing?

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,111 @@
+    {
+        $sessionUUID = $this->session->uuid;
+        $cookie = $this->cookieManager->get('uuid');
+        if (!(isset($cookie))) {
```
Minor style note: the extra parentheses around `(isset($cookie))` are not necessary in this context.

alexklbuckley replied:
Thanks @demiankatz this is added in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return string
+     */
+    public function setSecret()
```
I think setSecret should take `$cookieName` and `$secret` parameters, so specific values can be set if there is a desire to do so -- but $secret can default to null and generate a new uniqid when unspecified.

alexklbuckley replied:
Thanks @demiankatz I have added this in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SecureCookie.php:
```diff
@@ -0,0 +1,111 @@
+        $secret = uniqid();
+
+        // Clear then store secret in cookie
+        $this->cookieManager->clear('uuid');
```
Is it really necessary to clear before setting?

alexklbuckley replied:
thanks @demiankatz I agree that this probably is not necessary. I have removed clearing before setting from my follow-up

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley -- great progress, and this is looking pretty close to what I had in mind. See below for a few suggested refinements.

Also, it occurs to me that it might be useful to add something to the DevToolsController that can be used for testing permissions, and perhaps we could include the ability to set a specific secure cookie in that environment for testing/development purposes. Once this is ready for testing, I can work on building that if you think it's a good idea.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into secure_cookie_permission_provider

alexklbuckley committed:
Add secure cookie permission provider: Implement review feedback

- Check cookie and session container values based on $options value(s) in SecureCookie.php getPermissions().
- Remove redundant parentheses around (isset($cookie)).
- Pass $cookieName and $secret to SecureCookie.php setSecret() function.
- Remove clearing cookie value before setting secret in SecureCookie.php setSecret() function.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> Thanks, @alexklbuckley -- great progress, and this is looking pretty close to what I had in mind. See below for a few suggested refinements.
> 
> Also, it occurs to me that it might be useful to add something to the DevToolsController that can be used for testing permissions, and perhaps we could include the ability to set a specific secure cookie in that environment for testing/development purposes. Once this is ready for testing, I can work on building that if you think it's a good idea.

Thanks very much for all your feedback @demiankatz !

I've just pushed a follow-up which, I hope, addresses all your points.

Yes, I think adding something to the DevToolsController could be a good idea. Thanks for suggesting that and offering to work on building it!!

---

One more question, would you mind please advising what custom changes do you envisage would be required on a VuFind instance to use this SecureCookie permission provider?

For context, we will use this SecureCookie permission provider for AUT, in the scenario you discussed with Craig on this thread https://sourceforge.net/p/vufind/mailman/message/58783105/

I.e. I imagine we would need to create a custom controller to add an entry URL that calls SecureCookie.php setSecret() and sets the secret in the cookie and session container. Is that correct?

demiankatz commented:
Thanks, @alexklbuckley! It's a holiday here today so I won't have time to look at this too closely before I return to work tomorrow, but I'll try to work on my proposed DevToolsController functionality soon (I can push it directly here if you don't mind).

Regarding what needs to be customized, I think you have it exactly right: you just need a custom controller action to set the cookie. Part of the reason I want to create the DevToolsController code is that it should serve as an example of how to do this, so if others want to do something similar, we can show them how it can be done.

However, I have to confess, on thinking this through a little more, it does raise one big question: is there actually any point to this? What we're doing is setting values in a cookie and in the session and then making sure that they match. But we could just as easily have a permission provider that checks the session for a particular value, and it would have more or less the same effect. The "secure cookie" we are creating doesn't really add anything beyond storing the secret session value in a redundant place and blocking permission if people tamper with it. Beyond that, the existing session cookie already kind of serves the purpose of identifying the user. I'm trying to decide whether there are any edge cases where this mechanism would actually be useful -- one example I can imagine is that an external system generates a cookie, and you build a controller access that algorithmically validates the cookie value and pushes it into the session if it's valid but doesn't if it's not. I'm not sure if that's a realistic scenario, and even if it were realistic, it might not be very secure (if you can algorithmically validate a cookie, then somebody can algorithmically spoof it). But maybe there's some edge where this makes sense. But if we're just generating random values, maybe my original idea was just kind of silly.

Sorry to only be thinking of this now -- but I may have gotten so caught up in solving one aspect of the problem that I didn't really think through whether the problem was worth solving at all. ;-)

I'll be interested to hear what you think! And regardless of that, I think creating a DevToolsController tool to experiment with permissions is a good idea, so I'll go forward with that piece one way or another.

alexklbuckley commented:
> Thanks, @alexklbuckley! It's a holiday here today so I won't have time to look at this too closely before I return to work tomorrow, but I'll try to work on my proposed DevToolsController functionality soon (I can push it directly here if you don't mind).
> 
> Regarding what needs to be customized, I think you have it exactly right: you just need a custom controller action to set the cookie. Part of the reason I want to create the DevToolsController code is that it should serve as an example of how to do this, so if others want to do something similar, we can show them how it can be done.
> 
> However, I have to confess, on thinking this through a little more, it does raise one big question: is there actually any point to this? What we're doing is setting values in a cookie and in the session and then making sure that they match. But we could just as easily have a permission provider that checks the session for a particular value, and it would have more or less the same effect. The "secure cookie" we are creating doesn't really add anything beyond storing the secret session value in a redundant place and blocking permission if people tamper with it. Beyond that, the existing session cookie already kind of serves the purpose of identifying the user. I'm trying to decide whether there are any edge cases where this mechanism would actually be useful -- one example I can imagine is that an external system generates a cookie, and you build a controller access that algorithmically validates the cookie value and pushes it into the session if it's valid but doesn't if it's not. I'm not sure if that's a realistic scenario, and even if it were realistic, it might not be very secure (if you can algorithmically validate a cookie, then somebody can algorithmically spoof it). But maybe there's some edge where this makes sense. But if we're just generating random values, maybe my original idea was just kind of silly.
> 
> Sorry to only be thinking of this now -- but I may have gotten so caught up in solving one aspect of the problem that I didn't really think through whether the problem was worth solving at all. ;-)
> 
> I'll be interested to hear what you think! And regardless of that, I think creating a DevToolsController tool to experiment with permissions is a good idea, so I'll go forward with that piece one way or another.


Thanks so much for your reply @demiankatz 

Great idea about doing the DevToolsController functionality, so it serves as an example. Yes, I'm more than happy for you to push it directly here :)

Good point about if the securecookie permission provider is necessary.

If I'm understanding correctly, it sounds like the alternative would be creating a secure session factory and permission provider. That would provide a public helper function for setting a random secret (like uuid) in a session value.

I guess my question is, if we are not comparing a cookie value to a session value how would we go about validating that a session random value had not been tampered with?
i.e. what would the getPermissions() function in the permission provider check?

alexklbuckley commented:
If you could please let me know, if my understanding of the alternative approach (in https://github.com/vufind-org/vufind/pull/3882#issuecomment-2325561201 ), is correct then I can get onto making a secure session permission provider and factory :)

demiankatz commented:
@alexklbuckley, if you agree there's no compelling reason to involve a cookie here, I think the simplest solution is just to rename SecureCookie to SessionKey and remove the cookie-specific logic. The rest can stay mostly the same, though there's no reason to set a random value in the session container; it could just be a boolean true or something like that. Then you end up with configuring the name of the key to use in permissions.ini, using a helper method to actually set the value in the session container, and having the validation routine check whether the key from the configuration file has been set in the session container.

The user can't tamper with session values because they reside only on the server and are not exposed to the user's browser (unlike cookies, which can be directly edited by the user). Obviously, this is still potentially vulnerable to session hijacking attacks, and if the local custom mechanism to set the session is not properly secured, that could introduce problems as well, but those things are to be expected. I think the basic concept is sound. And then even though we don't have a SecureCookie method, we could note in the InsecureCookie documentation that SessionKey is a more secure alternative.

Does that make sense?

demiankatz commented:
@alexklbuckley, I have just opened #3899 to introduce a permission summary development tool -- it seemed better to keep that work separate from this work. Depending on which PR gets merged first, I will add the integration between the two parts in whichever PR remains open after the first part is merged.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into secure_cookie_permission_provider

alexklbuckley commented:
> @alexklbuckley, if you agree there's no compelling reason to involve a cookie here, I think the simplest solution is just to rename SecureCookie to SessionKey and remove the cookie-specific logic. The rest can stay mostly the same, though there's no reason to set a random value in the session container; it could just be a boolean true or something like that. Then you end up with configuring the name of the key to use in permissions.ini, using a helper method to actually set the value in the session container, and having the validation routine check whether the key from the configuration file has been set in the session container.
> 
> The user can't tamper with session values because they reside only on the server and are not exposed to the user's browser (unlike cookies, which can be directly edited by the user). Obviously, this is still potentially vulnerable to session hijacking attacks, and if the local custom mechanism to set the session is not properly secured, that could introduce problems as well, but those things are to be expected. I think the basic concept is sound. And then even though we don't have a SecureCookie method, we could note in the InsecureCookie documentation that SessionKey is a more secure alternative.
> 
> Does that make sense?

Thanks so much for helping clarify this for me! I really appreciate it.

Yes, that does make perfect sense. I am currently working on making those changes, switching SecureCookie to SessionKey.

I will push those changes to this PR shortly.

alexklbuckley committed:
Add sessionKey permission provider: Implement review feedback

- Switch SecureCookie.php and SecureCookieFactory.php to SessionKey.php and SessionKeyFactory.php
- Remove cookie-specific logic from SessionKey.php and SessionKeyFactory.php
- Switch helper method to set a boolean true value for the session key name (defined in permissions.ini)
- Make getPermissions() validation routine check whether the key from the permissions.ini file has been set
- Note in the InsecureCookie permissions.ini documentation that SessionKey is a more secure alternative.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> @alexklbuckley, if you agree there's no compelling reason to involve a cookie here, I think the simplest solution is just to rename SecureCookie to SessionKey and remove the cookie-specific logic. The rest can stay mostly the same, though there's no reason to set a random value in the session container; it could just be a boolean true or something like that. Then you end up with configuring the name of the key to use in permissions.ini, using a helper method to actually set the value in the session container, and having the validation routine check whether the key from the configuration file has been set in the session container.
> 
> The user can't tamper with session values because they reside only on the server and are not exposed to the user's browser (unlike cookies, which can be directly edited by the user). Obviously, this is still potentially vulnerable to session hijacking attacks, and if the local custom mechanism to set the session is not properly secured, that could introduce problems as well, but those things are to be expected. I think the basic concept is sound. And then even though we don't have a SecureCookie method, we could note in the InsecureCookie documentation that SessionKey is a more secure alternative.
> 
> Does that make sense?

Thanks so much @demiankatz ! I've pushed a follow-up implementing all of these changes. Please do let me know if there are any other changes you would like me to make? :)

alexklbuckley commented:
Hi @demiankatz I've been reading through https://vufind.org/wiki/development:plugins:controllers 

I'd just like to double check please, for the functionality we require in  https://sourceforge.net/p/vufind/mailman/message/58783105/  am I correct that I need to:

1. Create a new module - named, for example, vufind/module/CombinedSearch - like https://vufind.org/wiki/development:architecture:customizing_vufind#modules ?
2.  Create a new controller - like https://vufind.org/wiki/development:plugins:controllers#creating_a_new_controller ?
3. Then create a static route to my new controller?

demiankatz commented:
Alex,

Yes, that's the basic idea. It's probably easiest if you create a single module for all of your local customizations (for example, we use a VuFindVillanova module) rather than creating separate modules for separate features. A more granular approach is certainly a possibility, but in practice I think it might be more work to manage, and as long as the code is well-organized within your local module, it shouldn't make much difference.

You might find this video helpful; it goes through the process of setting up a controller step by step with an example:

https://vufind.org/wiki/videos:code_generators_2

- Demian

From: Alex Buckley ***@***.***>
Sent: Wednesday, September 4, 2024 4:49 AM
To: vufind-org/vufind ***@***.***>
Cc: Demian Katz ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [vufind-org/vufind] Add session key permission provider (PR #3882)


Hi @demiankatz<https://github.com/demiankatz> I've been reading through https://vufind.org/wiki/development:plugins:controllers

I'd just like to double check please, for the functionality we require in https://sourceforge.net/p/vufind/mailman/message/58783105/ am I correct that I need to:

  1.  Create a new module - named, for example, vufind/module/CombinedSearch - like https://vufind.org/wiki/development:architecture:customizing_vufind#modules ?
  2.  Create a new controller - like https://vufind.org/wiki/development:plugins:controllers#creating_a_new_controller ?
  3.  Then create a static route to my new controller?

-
Reply to this email directly, view it on GitHub<https://github.com/vufind-org/vufind/pull/3882#issuecomment-2328278492>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AACLOTKS4JYVKDM5TL3DO4TZU3CPDAVCNFSM6AAAAABNDIGTMCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMRYGI3TQNBZGI>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -50,7 +50,8 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
         'ipRegEx' => IpRegEx::class,
         'role' => Role::class,
-        'serverParam' => ServerParam::class,
+    'serverParam' => ServerParam::class,
```
There still seems to be some kind of whitespace setting problem somewhere that's causing indentation to get lost. Can you check your editor settings?

alexklbuckley replied:
thanks @demiankatz I think I've found the offending setting in my editor and have been able to see and fix this up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -66,7 +67,8 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
         IpRegEx::class => IpRegExFactory::class,
         Role::class => \Laminas\ServiceManager\Factory\InvokableFactory::class,
-        ServerParam::class => InjectRequestFactory::class,
+    ServerParam::class => InjectRequestFactory::class,
```
More indentation inconsistency.

alexklbuckley replied:
thanks @demiankatz I've included the fix for this in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+     * @param $session Session container
+     */
+    public function __construct(
+        Container $session
```
You can make this more concise by saying `protected Container $session` here, deleting the `protected $session;` definition up above, and removing the `$this->session = $session;` line below. This is called "constructor property promotion" -- it's a relatively new PHP feature designed to make constructors and related logic less verbose, and I'm becoming a big fan of it. :-)

alexklbuckley replied:
Wonderful tip, thanks @demiankatz . Very helpful to know about this great new PHP feature. I've pushed a follow-up making these changes!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+    {
+        foreach ((array)$options as $sessionKey) {
+            $this->debug("getPermissions: option sessionKey '{$sessionKey}'");
+            $sessionValue = $this->session->$sessionKey;
```
I think you could simplify most of this logic down to:

```
if (!($this->session->$sessionKey ?? false)) {
    $this->debug('getPermissions: result = false');
    return [];
}
```

I don't think there's an advantage to having separate clauses for the unset/false cases, especially if you show the same debug message in both scenarios anyway. :-)

alexklbuckley replied:
thanks @demiankatz I have included this in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * Set a boolean true value for a key in the Session container.
+     *
+     * @param $sessionKey - Set a boolean true value for this session key.
```
This is missing the type -- I think `string` is what you need.

alexklbuckley replied:
Thanks for spotting that! I've included a fix in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @param $sessionKey - Set a boolean true value for this session key.
+     *
+     * @return string
```
Since the actual value being stored is just a flag, there's no longer a need to return a value -- I think you can safely make this a `void` function now.

alexklbuckley replied:
thanks, this is included in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return string
+     */
+    public function setSessionValue($sessionKey)
```
I'd suggest using a more formally typed method definition:

```
public function setSessionValue(string $sessionKey): void
```

This will ensure that an appropriate input is provided, and then you can delete your lines 100-104 since it will be impossible to provide an invalid value.

alexklbuckley replied:
Thanks, this is included in my follow-up.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/SessionKey.php:
```diff
@@ -0,0 +1,111 @@
+        // Store boolean true value for the sessionKey
+        $this->session->$sessionKey = true;
+
+        return $this->session->$sessionKey;
```
You can also delete the return statement.

alexklbuckley replied:
thanks this is included in my follow-up.

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley, this looks good to me -- just see below for some suggested simplifications and minor fixes.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into secure_cookie_permission_provider

alexklbuckley committed:
Add sessionKey permission provider: Implement review feedback

- Fix incorrect indentation.
- Simplify session validation logic.
- Adding missing parameter type.
- Make setSessionValue() function a void function.
- Add a more formally typed method definition for setSessionValue()

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Add sessionKey permission provider: Implement review feedback

- Implement constructor property promotion to SessionKey.php to make the constructor function and and it's related logic less verbose.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> Alex, Yes, that's the basic idea. It's probably easiest if you create a single module for all of your local customizations (for example, we use a VuFindVillanova module) rather than creating separate modules for separate features. A more granular approach is certainly a possibility, but in practice I think it might be more work to manage, and as long as the code is well-organized within your local module, it shouldn't make much difference. You might find this video helpful; it goes through the process of setting up a controller step by step with an example: https://vufind.org/wiki/videos:code_generators_2 - Demian From: Alex Buckley ***@***.***> Sent: Wednesday, September 4, 2024 4:49 AM To: vufind-org/vufind ***@***.***> Cc: Demian Katz ***@***.***>; Mention ***@***.***> Subject: [EXTERNAL] Re: [vufind-org/vufind] Add session key permission provider (PR #3882) Hi @demiankatz<https://github.com/demiankatz> I've been reading through https://vufind.org/wiki/development:plugins:controllers I'd just like to double check please, for the functionality we require in https://sourceforge.net/p/vufind/mailman/message/58783105/ am I correct that I need to: 1. Create a new module - named, for example, vufind/module/CombinedSearch - like https://vufind.org/wiki/development:architecture:customizing_vufind#modules ? 2. Create a new controller - like https://vufind.org/wiki/development:plugins:controllers#creating_a_new_controller ? 3. Then create a static route to my new controller? - Reply to this email directly, view it on GitHub<[#3882 (comment)](https://github.com/vufind-org/vufind/pull/3882#issuecomment-2328278492)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AACLOTKS4JYVKDM5TL3DO4TZU3CPDAVCNFSM6AAAAABNDIGTMCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMRYGI3TQNBZGI>. You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

Thanks so much @demiankatz ! I will have a watch of that tutorial and have a go with implementing this.

alexklbuckley commented:
Hi @demiankatz thanks again for suggesting the tutorial I went through it and it was helpful.

I have made a new module (VuFindAUT) and controller in it, using the code generators.

![Screenshot from 2024-09-05 19-06-21](https://github.com/user-attachments/assets/ebe01704-3c93-4053-a19a-907d0749c55e)

Of course, right now it simply provides a home route for https://librarysearchdev.aut.ac.nz/vufind/VuFindAUT - like the below screenshot

![Screenshot from 2024-09-05 19-08-09](https://github.com/user-attachments/assets/e491f1ef-c8d6-4cac-b0e1-4e7a3cba2c07)

I hope it's alright for me to ask, what would you recommend I need to do to make it : "add an entry URL that sets the cookie (session value now!) and then redirect into the rest of the system." (from https://sourceforge.net/p/vufind/mailman/message/58783105/ )?

I really appreciate any assistance you can provide with this!

demiankatz committed:
Merge branch 'dev' into secure_cookie_permission_provider

demiankatz committed:
Add method for unsetting keys.

demiankatz committed:
Fix typo.

demiankatz committed:
Give permission provider plugin manager its own factory.

demiankatz committed:
Add sessionKey demo to devtool.

demiankatz committed:
Expand .ini comments.

demiankatz commented on module/VuFind/src/VuFind/Role/DynamicRoleProviderFactory.php:
```diff
@@ -49,60 +49,40 @@ class DynamicRoleProviderFactory implements FactoryInterface
      *
-     * @param ContainerInterface $serviceLocator Service locator
-     * @param array              $rbacConfig     LmcRbacMvc configuration
+     * @param ContainerInterface $container  Service container
```
While I was changing the code anyway, I renamed this argument to eliminate outdated terminology.

alexklbuckley replied:
Thanks @demiankatz !

demiankatz committed:
Remove unused code.

demiankatz committed:
Add missing newlines.

demiankatz reviewed (approved):
@alexklbuckley, as promised, I have updated the DevToolsController to include a demo of setting/unsetting session keys, so there's some existing code in the project showing how this works.

It was a little more complicated than I had anticipated because the plugin manager for permission providers was not easily accessible. I had to break apart some factory code to make it possible to directly retrieve the necessary permission provider. Now that that is done, though, it's pretty easy, and you can look at my recent changes to the DevToolsController here for an example of how to set a session key and then redirect to another page. I think you'll likely want to do something similar to that in your own custom controller.

I'm approving this now, and I think it's ready to merge, but I'll wait until you've had time to review my recent changes and try this for yourself before doing the actual merge.

Please let me know if anything remains unclear or if you run into trouble!

(Note also that because of my configuration changes here, you may need to `sudo rm -rf $VUFIND_LOCAL_DIR/cache/configs` after pulling my latest changes to ensure that old cached configs don't cause trouble).

alexklbuckley commented:
> @alexklbuckley, as promised, I have updated the DevToolsController to include a demo of setting/unsetting session keys, so there's some existing code in the project showing how this works.
> 
> It was a little more complicated than I had anticipated because the plugin manager for permission providers was not easily accessible. I had to break apart some factory code to make it possible to directly retrieve the necessary permission provider. Now that that is done, though, it's pretty easy, and you can look at my recent changes to the DevToolsController here for an example of how to set a session key and then redirect to another page. I think you'll likely want to do something similar to that in your own custom controller.
> 
> I'm approving this now, and I think it's ready to merge, but I'll wait until you've had time to review my recent changes and try this for yourself before doing the actual merge.
> 
> Please let me know if anything remains unclear or if you run into trouble!
> 
> (Note also that because of my configuration changes here, you may need to `sudo rm -rf $VUFIND_LOCAL_DIR/cache/configs` after pulling my latest changes to ensure that old cached configs don't cause trouble).

Thank you for all your work on this @demiankatz !

I've reviewed your changes and tested and it all looks to be working for me. 

I've also used the work you added in the DevToolsController as an example of how to set a session key and re-direct to another page. Using that I have my VuFindAUTController setting the session key and then redirecting back to the VuFind home page.

I've added sessionKey to the permissions.ini like so:

```
[SessionBasedPermission.EDSModule1]
sessionKey = aut_key
permission = access.EDSExtendedResults
```

I've done some local testing and have been able to correctly re-direct to login when I did not have the aut_key set.

Can I please check, am I right in thinking the permission = access.EDSExtendedResults permission gives access to all EDS results, or are you just treated as a guest who can see some EDS results but not all?

demiankatz commented:
> I've reviewed your changes and tested and it all looks to be working for me.

Great, thanks! I will go ahead and merge this now.

> Can I please check, am I right in thinking the permission = access.EDSExtendedResults permission gives access to all EDS results, or are you just treated as a guest who can see some EDS results but not all?

You are correct: access.EDSExtendedResults gives full access to EDS results; it sets the flag indicating that an authenticated user is accessing the API. Guest mode is used when the permission is unset. If you want some intermediate behavior that displays more than guest mode but does not grant full access, customization of some sort will be necessary.

alexklbuckley commented:
> > I've reviewed your changes and tested and it all looks to be working for me.
> 
> Great, thanks! I will go ahead and merge this now.
> 
> > Can I please check, am I right in thinking the permission = access.EDSExtendedResults permission gives access to all EDS results, or are you just treated as a guest who can see some EDS results but not all?
> 
> You are correct: access.EDSExtendedResults gives full access to EDS results; it sets the flag indicating that an authenticated user is accessing the API. Guest mode is used when the permission is unset. If you want some intermediate behavior that displays more than guest mode but does not grant full access, customization of some sort will be necessary.

Thank you very much @demiankatz !!!

Thanks again for all your help, great to get this upstreamed.
